torch/csrc/PyInterpreter.cpp

```diff
@@ -555,6 +555,150 @@ static void set_tensor_attr_with_capsule(
+    // existing SymIntArrayRef's from an old .sym_size() call.
+    auto new_buffer_size = new_size;
+    if (!is_buffer_allocated && new_size <= 5) {
+      // This is the smallvector optimization
+      new_buffer_size = 5;
+      set_to_smallvector = true;
+    }
+    T* ptr = new T[new_buffer_size];
+    auto capsule =
+        py::capsule(ptr, [](void* p) { delete[] reinterpret_cast<T*>(p); });
+    int64_t idx = 0;
+    for (auto it = obj.begin(); it != obj.end(); ++it, ++idx) {
+      ptr[idx] = py::cast<T>(*it);
+    }
+    // Set the buffer
+    set_tensor_attr_with_capsule(tensor, capsule, base_attr_name);
+    // Set the len buffer
+    py::handle(mb_obj.value()).attr(buffer_len_attr_name.c_str()) = new_size;
+    // Set whether or not the buffer is now using the smallvector optimization
+    // (sized to 5)
+    py::handle(mb_obj.value()).attr(buffer_issmallvector_attr_name.c_str()) =
+        set_to_smallvector;
+  } else {
+    auto curr_buffer_pyobj = py::handle(mb_obj.value()).attr(base_attr_name);
+    void* buffer_pycapsule =
+        PyCapsule_GetPointer(curr_buffer_pyobj.ptr(), nullptr);
+    auto curr_buffer = reinterpret_cast<T*>(buffer_pycapsule);
+
+    // Overwrite the buffer with our new values, but only if any of them changed
+    // (due to a metadata mutation).
+    // This is technically not thread safe (maybe we should put the GIL here?),
+    // but *only* if we actually have to update, which only happens if there's a
+    // metadata mutation.
+    int64_t idx = 0;
+    for (auto it = obj.begin(); it != obj.end(); ++it, ++idx) {
+      auto actual_val = py::cast<T>(*it);
+      if constexpr (std::is_same_v<T, c10::SymInt>) {
+        // if our SymInts are symbolic, we are *not* doing an equality check on
+        // the symints. we just want to see if the nodes are the same. this is
+        // because we don't want to introduce any guards here.
+        if (curr_buffer[idx].is_heap_allocated() !=
+            actual_val.is_heap_allocated()) {
+          curr_buffer[idx] = actual_val;
+        } else if (
+            !curr_buffer[idx].is_heap_allocated() &&
+            curr_buffer[idx] != actual_val) {
+          curr_buffer[idx] = actual_val;
+        } else if (
+            curr_buffer[idx].is_heap_allocated() &&
+            curr_buffer[idx].toSymNodeImplUnowned() !=
+                actual_val.toSymNodeImplUnowned()) {
+          curr_buffer[idx] = actual_val;
+        }
+      } else {
+        if (curr_buffer[idx] != actual_val) {
+          curr_buffer[idx] = actual_val;
+        }
+      }
+    }
+  }
+
+  // The correct data is now stored at the buffer - read and return it.
+  auto curr_buffer_pyobj = py::handle(mb_obj.value()).attr(base_attr_name);
+  void* buffer_pycapsule =
+      PyCapsule_GetPointer(curr_buffer_pyobj.ptr(), nullptr);
+  auto curr_buffer = reinterpret_cast<T*>(buffer_pycapsule);
+  return c10::ArrayRef<T>(curr_buffer, new_size);
+}
```

Comment:
Meh. Here's a better idea: literally store a SizesAndStrides object inside the capsule, and then reuse the implementation there. Just need some placement-new/delete to get things setup initially.

Reply:
One reason that seems a bit painful is that `SizesAndStrides` stores the sizes and strides together in a single location in memory. The lazy updating of the buffer during metadata mutations makes that kind of painful here, because the flow in the PR as written is currently:

```
# For a normal tensor this will eagerly cause the SizesAndStrides buffer to do a reallocation
# But for subclasses, we have an independent size and stride buffer that will be lazily reallocated on first access
x.resize_(...)
...
# For subclasses, at this point, we re-allocate the size buffer when we detect that our buffer is the wrong size
sizes = x.size()

# For subclasses, at this point, we re-allocate the stride buffer when we detect that our buffer is the wrong size
sizes = x.size()
```

Reply:
But using a SizeAndStride object here seems to imply that we should actually store the sizes and strides together into that single object. That would mean that when a user calls `sym_size()` and we query the subclass for the updated size, we would also need to query for the updated stride, so we can stuff them both into the `SizeAndStride` object that we're using.

Is that what you were thinking? Or were you more just thinking of only re-using `SizeAndStrides` for the SmallVector optimization, and not to actually store the two of them together.

Reply:
Hmm. Yes, I think I am concretely proposing you force both a size and stride query whenever one query occurs.

Reply:
ok now I am tempted to try to root-cause the bug that required me to add the SmallVector optimization :) Failing that, I'll do the SizeAndStride update.

(Actually, double checking - if we can avoid needing to make the SmallVector optimization happen here, are you ok with a vanilla `new T[updated_size]` call?)

Reply:
Yeah, if you don't need to exactly mimic SizeAndStrides, just something simple is good. But if you DO want to mimic SizeAndStrides, might as well do the real thing.

By the way, you could avoid calling the stride query if you don't resize.

Reply:
aaand... tests still pass when I remove the small vector optimization. I thought I checked this before I added it, but I must have also fixed another bug at the same time or something...

PR updated to not do the SmallVector thing and just do `new T[updated_size]`.